datasource/src/main/scala/quasar/blobstore/azure/rx.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.blobstore.azure
+
+import slamdata.Predef._
+
+import cats.effect._
+import cats.implicits._
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.numeric.Positive
+import fs2.{RaiseThrowable, Stream}
+import fs2.concurrent.Queue
+import io.reactivex.{Flowable, Single, SingleObserver}
+import io.reactivex.observers.DisposableSingleObserver
+import io.reactivex.subscribers.DefaultSubscriber
+
+object rx {
+  final class AsyncSubscriber[F[_]: Sync, A](cb: Either[Throwable, Option[F[A]]] => Unit) extends DefaultSubscriber[A] {
+
+    override def onStart(): Unit =
+      request(1)
+
+    def onNext(a: A): Unit = {
+      cb(Right(Some(Sync[F].delay(request(1)).as(a))))
+      request(1)
+    }
+
+    def onError(t: Throwable): Unit =
+      cb(Left(t))
+
+    def onComplete: Unit =
+      cb(Right(none))
+  }
+
+  final class AsyncObserver[A] extends DisposableSingleObserver[A] {
+
+    @SuppressWarnings(Array("org.wartremover.warts.Null", "org.wartremover.warts.Var"))
+    private var callback: Either[Throwable, A] => Unit = _
+
+    def setCallback(cb: Either[Throwable, A] => Unit): Unit = {
+      this.callback = cb
+    }
+
+    override def onStart(): Unit = ()
+
+    override def onSuccess(a: A) =
+      if (callback != null) callback(Right(a))
+      else ()
+
+    override def onError(t: Throwable) =
+      if (callback != null) callback(Left(t))
+      else ()
+  }
+
+  def flowableToStream[F[_]: ConcurrentEffect: RaiseThrowable, A](
+      f: Flowable[A],
+      maxQueueSize: Int Refined Positive): Stream[F, A] =
+    handlerToStream(flowableToHandler(f), maxQueueSize)
+
+  def flowableToHandler[F[_]: Sync, A](flowable: Flowable[A]): (Either[Throwable, Option[F[A]]] => Unit) => Unit = { cb =>
```

Comment:
The return type here should be `(Either[Throwable, Option[F[A]]] => Unit) => F[Unit]` so we capture the side-effect of subscription.